tests/trunner.nim

```diff
@@ -19,28 +19,45 @@ proc runCmd(file, options = ""): auto =
     echo result[0]
     echo result[1]
 
-proc testCodegenStaticAssert() =
-  let (output, exitCode) = runCmd("ccgbugs/mstatic_assert.nim")
-  doAssert "sizeof(bool) == 2" in output
-  doAssert exitCode != 0
-
-proc testCTFFI() =
-  let (output, exitCode) = runCmd("vm/mevalffi.nim", "--experimental:compiletimeFFI")
-  let expected = """
-hello world stderr
-hi stderr
-foo
-foo:100
-foo:101
-foo:102:103
-foo:102:103:104
-foo:0.03:asdf:103:105
-ret={s1:foobar s2:foobar age:25 pi:3.14}
-"""
-  doAssert output == expected, output
-  doAssert exitCode == 0
-
 when defined(nimHasLibFFIEnabled):
-  testCTFFI()
+  block: # mevalffi
+    let (output, exitCode) = runCmd("vm/mevalffi.nim", "--experimental:compiletimeFFI")
+    let expected = """
+  hello world stderr
+  hi stderr
+  foo
+  foo:100
+  foo:101
+  foo:102:103
+  foo:102:103:104
+  foo:0.03:asdf:103:105
+  ret={s1:foobar s2:foobar age:25 pi:3.14}
+  """
+    doAssert output == expected, output
+    doAssert exitCode == 0
+
 else: # don't run twice the same test
-  testCodegenStaticAssert()
+  block: # mstatic_assert
+    let (output, exitCode) = runCmd("ccgbugs/mstatic_assert.nim", "-d:caseBad")
+    doAssert "sizeof(bool) == 2" in output, output
+    doAssert exitCode != 0
+
+  block: # mabi_check
+    let (output, exitCode) = runCmd("ccgbugs/mabi_check.nim")
+    # on platforms that support _StaticAssert natively, will show full context, eg:
+    # error: static_assert failed due to requirement 'sizeof(unsigned char) == 8'
+    # "backend & Nim disagree on size for: BadImportcType{int64} [declared in mabi_check.nim(1, 6)]"
+    doAssert "sizeof(unsigned char) == " in output, output
+    doAssert exitCode != 0
+
+  block: # mimportc_size_check
+    block:
+      let (_, exitCode) = runCmd("ccgbugs/mimportc_size_check.nim")
+      doAssert exitCode == 0
+    block:
+      let (output, exitCode) = runCmd("ccgbugs/mimportc_size_check.nim", "-d:caseBad")
+      doAssert "sizeof(struct Foo2) == 1" in output, output
+      doAssert "sizeof(Foo5) == 16" in output, output
+      doAssert "sizeof(Foo5) == 3" in output, output
+      doAssert "sizeof(struct Foo6) == " in output, output
+      doAssert exitCode != 0
```

Comment:
I don't know this file. It somehow does not exist in my Nim repository. But code generation checks should be done with the ``ccodeCheck``. You can look at [talign.nim](https://github.com/nim-lang/Nim/blob/devel/tests/align/talign.nim) for an example. You have to consult @Araq for its syntax though. I don't know of any documentation about it, nor do I really understand it.

Reply:
introduced in https://github.com/nim-lang/Nim/pull/13798 but indirectly existed since CT FFI. See rationale here https://github.com/nim-lang/Nim/pull/13798#issuecomment-605838197
> I don't know of any documentation about it, nor do I really understand it.

exactly. testament spec, being declarative, is good for simple things but, just like `nim.cfg`, not suitable everywhere, and shouldn't be turned into something turing complete when we can already simply run nim code which everyone will be familiar with.